Sources/TuistGenerator/GraphMappers/AutogeneratedSchemesGraphMapper.swift

```diff
@@ -0,0 +1,65 @@
+import Foundation
+import TuistCore
+
+/// A graph mapper that auto-generates schemes for each of the targets of the graph
+/// if the user hasn't already defined schemes for those.
+public class AutogeneratedSchemesGraphMapper: GraphMapping {
+    // MARK: - Init
+
+    public init() {}
+
+    // MARK: - GraphMapping
+
+    public func map(graph: Graph) throws -> (Graph, [SideEffectDescriptor]) {
+        let projects = graph.projects.map { (project: Project) -> Project in
+            let schemeNames = Set(project.schemes.map { $0.name })
+            let schemes = project.schemes
+
+            let autogeneratedSchemes = project.targets.compactMap { (target: Target) -> Scheme? in
+                let scheme = self.createDefaultScheme(target: target,
+                                                      project: project,
+                                                      buildConfiguration: project.defaultDebugBuildConfigurationName,
+                                                      graph: graph)
+                // The user has already defined a scheme with that name.
+                if schemeNames.contains(scheme.name) { return nil }
+                return scheme
+            }
+
+            return project.with(schemes: schemes + autogeneratedSchemes)
+        }
+
+        let mappedGraph = Graph(name: graph.name,
+                                entryPath: graph.entryPath,
+                                entryNodes: graph.entryNodes,
+                                projects: projects,
+                                cocoapods: graph.cocoapods,
+                                packages: graph.packages,
+                                precompiled: graph.precompiled,
+                                targets: graph.targets)
```

Comment:
`graph.targets` is an `[AbsolutePath: TargetNode]`, and `TargetNode` has a `project` reference within it. After this mapping `mappedGraph.projects` will have different projects instances than `mappedGraph.targets.values.flatMap { $0 }.map { $0.project }`. 



![Screen Shot 2020-05-19 at 6 58 09 AM](https://user-images.githubusercontent.com/11914919/82290204-42a1bd80-999e-11ea-9b6e-cd6341bfd927.png)

This is a very similar to the issues we saw in #1206 - the `Graph` we're dealing with here is a cache optimized for lookups and isn't suitable for mapping due to the recurrence of the same elements in different containers. 

At the moment it so happens there aren't any inconsistencies and hence why it works as expected, but having different projects instances in the graph can affect the behavior of other mappers or the generation process.

Reply:
Ouch, I missed this one entirely 🤦‍♂️. I'm so happy that we are moving away from using crossed references like this one.